File: app/src/main/java/ru/rewindforce/concerts/ConcertDetails/ConcertDetailsPresenter.java
package ru.rewindforce.concerts.ConcertDetails;

import java.util.ArrayList;

import okhttp3.MediaType;
import okhttp3.RequestBody;
import ru.rewindforce.concerts.HomeScreen.Band;

class ConcertDetailsPresenter {
    private ConcertDetailsFragment fragment;
    private ConcertDetailsModel model;

    ConcertDetailsPresenter() {
        model = new ConcertDetailsModel();
    }

    void attachFragment(ConcertDetailsFragment fragment) {
        this.fragment = fragment;
    }

    void detachFragment() {
        this.fragment = null;
    }

    void getStatus(String token, String uid, int id) {
        model.getStatus(token, uid, id, new ConcertDetailsModel.ConcertsCallback() {
            @Override
            public void onStatus(String status) {
                if (fragment != null) fragment.onStatusLoad(status);
            }

            @Override
            public void onError(boolean shouldLoadAgain) {
                if (fragment != null) fragment.onLoadError(shouldLoadAgain);
            }
        });
    }

    void getLineUp(int concert_id) {
        model.getLineUp(concert_id, new ConcertDetailsModel.ConcertsCallback() {
            @Override
            public void onLineUp(ArrayList<Band> bands) {
                if (fragment != null) {
                    fragment.removeDummyLineUp();
                    for (int i = 0; i < bands.size(); i++) fragment.onLineUpLoad(bands.get(i), i);
                }
            }

            @Override
            public void onError(boolean shouldLoadAgain) {
                if (fragment != null) fragment.onLoadError(shouldLoadAgain);
            }
        });
    }

    void putToWishlist(int button_id, String token, String uid, String state, int id) {
        RequestBody partUid = RequestBody.create(MediaType.parse("text/plain"), uid);
        RequestBody partState = RequestBody.create(MediaType.parse("text/plain"), state);
        RequestBody partId = RequestBody.create(MediaType.parse("text/plain"), String.valueOf(id));

        model.putToWishlist(button_id, token, partUid, partState, partId, new ConcertDetailsModel.ConcertsCallback() {
                @Override
                public void onResponse(int button_id) {
                    if (fragment != null) fragment.onConcertsLoad(button_id);
                }

                @Override
                public void onError(boolean shouldLoadAgain) {
                    if (fragment != null) fragment.onLoadError(shouldLoadAgain);
                }
            });
    }
}

File: app/src/main/java/ru/rewindforce/concerts/details/ConcertDetailsPresenter.kt
package ru.rewindforce.concerts.details

import okhttp3.MediaType
import okhttp3.RequestBody
import okhttp3.ResponseBody
import ru.rewindforce.concerts.data.Band
import ru.rewindforce.concerts.data.Comment
import ru.rewindforce.concerts.utils.Presenter
import java.util.ArrayList

class ConcertDetailsPresenter: Presenter<ConcertDetailsModel>() {
    init { this.model = ConcertDetailsModel() }

    fun getStatus(token: String, userUID: String, concertUID: String) {
        val args = mapOf("token" to token, "userUID" to userUID, "concertUID" to concertUID)
        model.getStatus(args, callback(
                onResponse = { _: Any, code: Int -> when (code) {
                    200 -> fragment?.onStatusLoad("going")
                    404 -> fragment?.onStatusLoad("not going")
                    204 -> fragment?.onError()
                }},
                onFailure = { fragment?.onError() }
        ))
    }

    fun getLineUp(bandsUID: ArrayList<String>) {
        val args = mapOf("bandsUID" to bandsUID)
        model.getLineUp(args, callback(
                onResponse = {body: ArrayList<Band>, _ -> fragment?.onLineUpLoad(body) },
                onFailure = { fragment?.onError() }
        ))
    }

    fun putToWishlist(buttonID: Int, token: String, userUID: String, state: String, concertUID: String) {
        val args = mapOf("token" to token, "authUID" to userUID, "concertUID" to concertUID)

        val callback = callback(onResponse = {_: Any, _ -> fragment?.onConcertsLoad(buttonID) },
                                onFailure = { fragment?.onError() })

        when (state) {
            "going", "maybe" -> model.putToWishlist(args, callback)
            "not going" -> model.deleteFromWishlist(args, callback)
        }
    }

    fun getComments(concertUID: String, limit: Int, offset: Int) {
        val args = mapOf("concertUID" to concertUID, "limit" to limit, "offset" to offset)
        model.getComments(args, callback(
                onResponse = {body: ArrayList<Comment>, _ -> fragment?.onCommentsLoad(body) },
                onFailure = { fragment?.onCommentsLoad(ArrayList()) }))
    }

    fun sendComment(concertUID: String, token: String, userUID: String, message: String) {
        val partMessage = RequestBody.create(MediaType.parse("text/plain"), message)
        val args = mapOf("concertUID" to concertUID, "token" to token, "userUID" to userUID, "message" to partMessage)
        model.sendComment(args, callback(
                onResponse = {body: Comment, _ -> fragment?.onCommentPosted(body) },
                onFailure = { fragment?.onCommentError() }))
    }

    fun likeComment(concertUID: String, commentUID: String, token: String, userUID: String, adapterPos: Int) {
        val args = mapOf("concertUID" to concertUID, "commentUID" to commentUID, "token" to token, "userUID" to userUID)
        model.likeComment(args, callback(
                onResponse = {_: Any, _ -> fragment?.onLiked(adapterPos) },
                onFailure = { fragment?.onError() }))
    }

    fun dislikeComment(concertUID: String, commentUID: String, token: String, userUID: String, adapterPos: Int) {
        val args = mapOf("concertUID" to concertUID, "commentUID" to commentUID, "token" to token, "userUID" to userUID)
        model.dislikeComment(args, callback(
                onResponse = {_: Any, _ -> fragment?.onDisliked(adapterPos) },
                onFailure = { fragment?.onError() }))
    }

    fun deleteComment(concertUID: String, commentUID: String, token: String, userUID: String, adapterPos: Int) {
        val args = mapOf("concertUID" to concertUID, "commentUID" to commentUID, "token" to token, "userUID" to userUID)
        model.deleteComment(args, callback(
                onResponse = {_: Any, _ -> fragment?.onCommentDeleted(adapterPos) },
                onFailure = { fragment?.onError() }))
    }

    fun editComment(message: String, concertUID: String, commentUID: String, token: String,
                    userUID: String, adapterPos: Int) {
        val partMessage = RequestBody.create(MediaType.parse("text/plain"), message)
        val args = mapOf("concertUID" to concertUID, "commentUID" to commentUID, "token" to token,
                         "userUID" to userUID, "message" to partMessage)
        model.editComment(args, callback(
                onResponse = {body: ResponseBody, _ -> fragment?.onCommentEdited(body.string(), adapterPos) },
                onFailure = { fragment?.onError() }))
    }
}